server/src/main/java/org/opensearch/cluster/routing/allocation/decider/NodeOverloadAwareAllocationDecider.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing.allocation.decider;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.routing.RoutingNode;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.settings.Setting.Property;
+
+import java.util.function.BiPredicate;
+
+/**
+ * This {@link NodeOverloadAwareAllocationDecider} controls shard over-allocation
+ * due to node failures or otherwise on the surviving nodes
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.provisioned_capacity: N
+ * </pre>
+ * <p>
+ * and prevent allocation on the surviving nodes of the under capacity cluster
+ * based on oveload factor defined as a percentage by
+ * <pre>
+ * cluster.routing.allocation.overload_awareness.factor: X
+ * </pre>
+ */
+public class NodeOverloadAwareAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "overload_awareness";
+
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING =
+        Setting.intSetting("cluster.routing.allocation.overload_awareness.provisioned_capacity", -1, -1,
+            Property.Dynamic, Property.NodeScope);
+    public static final Setting<Integer> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING =
+        Setting.intSetting("cluster.routing.allocation.overload_awareness.factor", 50, -1, Property.Dynamic,
+            Property.NodeScope);
+    public static final Setting<Boolean> CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_ALLOW_UNASSIGNED_PRIMARIES_SETTING =
+        Setting.boolSetting("cluster.routing.allocation.overload_awareness.allow_unassigned_primaries",
+            true, Setting.Property.Dynamic, Property.NodeScope);
+
+    private volatile int provisionedCapacity;
+
+    private volatile int overloadFactor;
+
+    private volatile boolean allowUnassignedPrimaries;
+
+    private static final Logger logger = LogManager.getLogger(NodeOverloadAwareAllocationDecider.class);
+
+    public NodeOverloadAwareAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.overloadFactor = CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING.get(settings);
+        this.provisionedCapacity = CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING.get(settings);
```

Comment:
Do you think an uber level setting for decider to be ``enabled/disabled`` makes sense? Given there are multiple scenarios when user would want to temporarily disable the feature (such as cluster sizing changes)?

Reply:
The overload limit can be set to -1 meaning unlimited, this is also how ShardLimitAllocationDecider does, avoiding too many settings here